src/app/clusters/operational-state-server/operational-state-delegate.h

```diff
@@ -0,0 +1,323 @@
+{
+    GenericOperationalError(uint8_t state, Optional<chip::CharSpan> label = NullOptional,
+                            Optional<chip::CharSpan> details = NullOptional)
+    {
+        Set(state, label, details);
+    }
+
+    GenericOperationalError(const GenericOperationalError & error) { *this = error; }
+
+    GenericOperationalError & operator=(const GenericOperationalError & error)
+    {
+        Set(error.errorStateID, error.errorStateLabel, error.errorStateDetails);
+        return *this;
+    }
+
+    void Set(uint8_t state, Optional<chip::CharSpan> label = NullOptional, Optional<chip::CharSpan> details = NullOptional)
+    {
+        errorStateID = state;
+        if (label.HasValue())
+        {
+            memset(mErrorStateLabelBuffer, 0, sizeof(mErrorStateLabelBuffer));
+            if (label.Value().size() > sizeof(mErrorStateLabelBuffer))
+            {
+                memcpy(mErrorStateLabelBuffer, label.Value().data(), sizeof(mErrorStateLabelBuffer));
+                errorStateLabel.SetValue(CharSpan(mErrorStateLabelBuffer, sizeof(mErrorStateLabelBuffer)));
+            }
+            else
+            {
+                memcpy(mErrorStateLabelBuffer, label.Value().data(), label.Value().size());
+                errorStateLabel.SetValue(CharSpan(mErrorStateLabelBuffer, label.Value().size()));
+            }
+        }
+        else
+        {
+            errorStateLabel = NullOptional;
+        }
+
+        if (details.HasValue())
+        {
+            memset(mErrorStateDetailsBuffer, 0, sizeof(mErrorStateDetailsBuffer));
+            if (details.Value().size() > sizeof(mErrorStateDetailsBuffer))
+            {
+                memcpy(mErrorStateDetailsBuffer, details.Value().data(), sizeof(mErrorStateDetailsBuffer));
+                errorStateDetails.SetValue(CharSpan(mErrorStateDetailsBuffer, sizeof(mErrorStateDetailsBuffer)));
+            }
+            else
+            {
+                memcpy(mErrorStateDetailsBuffer, details.Value().data(), details.Value().size());
+                errorStateDetails.SetValue(CharSpan(mErrorStateDetailsBuffer, details.Value().size()));
+            }
+        }
+        else
+        {
+            errorStateDetails = NullOptional;
+        }
+    }
+
+private:
+    char mErrorStateLabelBuffer[kOperationalErrorLabelMaxSize];
+    char mErrorStateDetailsBuffer[kOperationalErrorDetailsMaxSize];
+};
+
+/**
+ * A class which represents the operational phase of an Operational State cluster derivation instance.
+ */
+struct GenericOperationalPhase
+{
+    GenericOperationalPhase(app::DataModel::Nullable<CharSpan> name) { Set(name); }
+
+    GenericOperationalPhase(const GenericOperationalPhase & ph) { *this = ph; }
+
+    GenericOperationalPhase & operator=(const GenericOperationalPhase & ph)
+    {
+        Set(ph.mPhaseName);
+        return *this;
+    }
+
+    bool IsMissing() const { return mPhaseName.IsNull(); }
+    app::DataModel::Nullable<CharSpan> mPhaseName;
```

Comment:
Please:

1. File a followup to make this not nullable, since it can never be null in the spec.
2. Have a comment referencing this followup here.

This applies in general to things that are deferred to followups.

Reply:
![image](https://github.com/project-chip/connectedhomeip/assets/106149377/4e6be252-984a-4cfa-a77e-76a7c1a24eb2)
I have downloaded the lastest spec from https://github.com/CHIP-Specifications/connectedhomeip-spec/tree/build-sample, is seems that it can be null, so can you have a check? @bzbarsky-apple

Reply:
The cited code is not about PhaseList, though.  It's about a specific phase _in the list_.  Those can never be null.

Reply:
Put another way, the delegate should have some way to ask it whether PhaseList should be null.  If it's not, then GetOperationalPhaseAtIndex should just hand out phases.  It could probably take MutableCharSpan as its out param (initially pointing to a caller-provided buffer) without having a GenericOperationalPhase struct at all....